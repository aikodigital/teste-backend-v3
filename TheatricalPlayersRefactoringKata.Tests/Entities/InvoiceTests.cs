using ApprovalTests;
using ApprovalTests.Reporters;
using System;
using System.Collections.Generic;
using TheatricalPlayersRefactoringKata.Domain;
using TheatricalPlayersRefactoringKata.Domain.Enums;
using TheatricalPlayersRefactoringKata.Domain.Validation;
using Xunit;

namespace TheatricalPlayersRefactoringKata.Tests.Entities
{
    public class InvoiceTests : IDisposable
    {
        private readonly List<Performance> _performances;

        public InvoiceTests()
        {
            _performances = new List<Performance>
            {
                new(new Play("Hamlet", 4024, PlayTypeEnum.Tragedy), 55),
                new(new Play("As You Like It", 2670, PlayTypeEnum.Comedy), 35),
                new(new Play("Henry V", 3227, PlayTypeEnum.History),20)
            };
        }

        [Fact]
        public void CreateInvoiceWithValidValues()
        {
            var customer = "BigCo";

            Invoice newInvoice = new Invoice(customer, _performances);

            Assert.Equal(customer, newInvoice.Customer);
            Assert.Contains(newInvoice.Performances, (performance) => _performances.Contains(performance));
        }

        [Fact]
        public void CreateInvoiceWithInalidCustomer()
        {
            var exception = Assert.Throws<DomainExceptionValidation>(() => new Invoice("", _performances));

            Assert.Equal("Invalid customer. Customer is required.", exception.Message);
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void PrintInvoiceStatement()
        {
            var customer = "BigCo";

            Invoice newInvoice = new Invoice(customer, _performances);

            string invoiceStatement = newInvoice.PrintInvoiceStatement();

            Approvals.Verify(invoiceStatement);
        }

        public void Dispose()
        {
            _performances.Clear();
        }
    }
}
