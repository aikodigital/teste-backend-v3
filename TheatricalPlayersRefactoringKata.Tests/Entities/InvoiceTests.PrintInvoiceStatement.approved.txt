Statement for BigCo
  Hamlet: $650.00 (55 seats)
  As You Like It: $547.00 (35 seats)
  Henry V: $705.40 (20 seats)
Amount owed is $1,902.40
You earned 37 credits
